compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -669,10 +669,28 @@ trait Implicits { self: Typer =>
             case result: AmbiguousImplicits => true
             case _ => false
           }
+
+        // Is Eq[X, Null] or Eq[Null, X] where !(X <:< AnyVal)?
+        def eqNullable(tp1: Type, tp2: Type): Boolean = {
+          val other =
+            if (tp1.stripTypeVar eq defn.NullType) tp2
+            else if (tp2.stripTypeVar eq defn.NullType) tp1
+            else NoType
+
+          (other ne NoType) && !other.derivesFrom(defn.AnyValClass)
+        }
+
         def validEqAnyArgs(tp1: Type, tp2: Type) = {
           List(tp1, tp2).foreach(fullyDefinedType(_, "eqAny argument", pos))
-          assumedCanEqual(tp1, tp2) || !hasEq(tp1) && !hasEq(tp2) ||
-            { implicits.println(i"invalid eqAny[$tp1, $tp2]"); false }
```

Comment:
I'd leave it as it is and change assumedCanEqual instead, so that it returns true if one type is Null and the other is not a value type.